File: samples/repository_samples/javascript_crypto/crypto_service.js
/**
 * QNetra Sample: JavaScript Cryptographic Code
 * Tests Node.js crypto module, jsonwebtoken, and pattern matching.
 * Used as a test fixture for JavaScriptAnalyzer.
 */

'use strict';

const crypto = require('crypto');
const jwt = require('jsonwebtoken');
const CryptoJS = require('crypto-js');

// ============================================================
// AES Encryption — Node.js built-in crypto
// ============================================================

/**
 * AES-128-CBC encryption — Grover reduces to 64-bit quantum security.
 * @param {Buffer} data - Plaintext
 * @param {Buffer} key  - 16-byte (128-bit) key
 * @returns {Object} {iv, encrypted}
 */
function encryptAES128CBC(data, key) {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-128-cbc', key, iv);
    const encrypted = Buffer.concat([cipher.update(data), cipher.final()]);
    return { iv, encrypted };
}

/**
 * AES-256-GCM encryption — 128-bit post-quantum security (acceptable).
 */
function encryptAES256GCM(data, key) {
    const iv = crypto.randomBytes(12);
    const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
    const encrypted = Buffer.concat([cipher.update(data), cipher.final()]);
    const authTag = cipher.getAuthTag();
    return { iv, encrypted, authTag };
}

// ============================================================
// RSA Key Generation — Quantum Vulnerable
// ============================================================

async function generateRSAKeyPair() {
    return crypto.generateKeyPair('rsa', {
        modulusLength: 2048,  // 2048-bit RSA — fully broken by Shor's algorithm
        publicKeyEncoding: { type: 'pkcs1', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs1', format: 'pem' },
    });
}

// ============================================================
// Hashing
// ============================================================

function hashMD5(data) {
    // MD5 is CLASSICALLY BROKEN
    return crypto.createHash('md5').update(data).digest('hex');
}

function hashSHA256(data) {
    return crypto.createHash('sha256').update(data).digest('hex');
}

function hashSHA1(data) {
    // SHA-1 is CLASSICALLY BROKEN — SHAttered collision
    return crypto.createHash('sha1').update(data).digest('hex');
}

// ============================================================
// HMAC
// ============================================================

function createHMACSHA256(key, message) {
    return crypto.createHmac('sha256', key).update(message).digest('hex');
}

// ============================================================
// JWT — Common algorithm patterns
// ============================================================

function signJWT(payload, secret) {
    // HS256 = HMAC-SHA256 — symmetric, no quantum concern at 256-bit
    return jwt.sign(payload, secret, { algorithm: 'HS256', expiresIn: '1h' });
}

function signJWTWithRSA(payload, privateKey) {
    // RS256 = RSA-SHA256 — QUANTUM VULNERABLE (Shor's algorithm)
    return jwt.sign(payload, privateKey, { algorithm: 'RS256' });
}

// ============================================================
// PBKDF2
// ============================================================

function deriveKey(password, salt) {
    // PBKDF2-SHA256 with 310,000 iterations (OWASP 2023 recommendation)
    return crypto.pbkdf2Sync(password, salt, 310000, 32, 'sha256');
}

// ============================================================
// CryptoJS (browser/universal library)
// ============================================================

function encryptWithCryptoJS(message, key) {
    // AES in CBC mode (CryptoJS default)
    return CryptoJS.AES.encrypt(message, key).toString();
}

function computeMD5(message) {
    // MD5 — CLASSICALLY BROKEN
    return CryptoJS.MD5(message).toString();
}

module.exports = {
    encryptAES128CBC, encryptAES256GCM, generateRSAKeyPair,
    hashMD5, hashSHA256, hashSHA1, createHMACSHA256,
    signJWT, signJWTWithRSA, deriveKey, encryptWithCryptoJS, computeMD5,
};

File: scripts/java_crypto/CryptoService.java
/**
 * QNetra Sample: Java Cryptographic Code
 * Tests javax.crypto, java.security patterns and getInstance() calls.
 * Used as a test fixture for JavaAnalyzer.
 */

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.*;
import java.security.spec.ECGenParameterSpec;
import java.util.Base64;

public class CryptoService {

    // ================================================================
    // AES Encryption — Grover-Impacted
    // ================================================================

    public static byte[] encryptAES128GCM(byte[] plaintext, byte[] keyBytes) throws Exception {
        // AES-128-GCM — Grover's algorithm reduces to 64-bit quantum security
        SecretKey key = new SecretKeySpec(keyBytes, 0, 16, "AES");
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        GCMParameterSpec spec = new GCMParameterSpec(128, new byte[12]);
        cipher.init(Cipher.ENCRYPT_MODE, key, spec);
        return cipher.doFinal(plaintext);
    }

    public static SecretKey generateAES256Key() throws Exception {
        // AES-256 key generation — Grover reduces to 128-bit quantum security
        KeyGenerator keyGen = KeyGenerator.getInstance("AES");
        keyGen.init(256);  // 256-bit key
        return keyGen.generateKey();
    }

    // ================================================================
    // DES Encryption — Classically Broken
    // ================================================================

    public static byte[] encryptDES(byte[] plaintext, byte[] keyBytes) throws Exception {
        // DES is CLASSICALLY BROKEN — 56-bit key, do not use
        SecretKey key = new SecretKeySpec(keyBytes, "DES");
        Cipher cipher = Cipher.getInstance("DES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, key);
        return cipher.doFinal(plaintext);
    }

    // ================================================================
    // RSA Key Generation — Quantum Vulnerable
    // ================================================================

    public static KeyPair generateRSA2048() throws Exception {
        // RSA-2048 — QUANTUM VULNERABLE (Shor's algorithm)
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("RSA");
        kpg.initialize(2048);
        return kpg.generateKeyPair();
    }

    public static byte[] signWithRSA(PrivateKey privateKey, byte[] data) throws Exception {
        // RSA-SHA256 signature — QUANTUM VULNERABLE
        Signature sig = Signature.getInstance("SHA256withRSA");
        sig.initSign(privateKey);
        sig.update(data);
        return sig.sign();
    }

    // ================================================================
    // ECDSA — Quantum Vulnerable
    // ================================================================

    public static KeyPair generateECDSA() throws Exception {
        // ECDSA on secp256r1 — QUANTUM VULNERABLE
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("EC");
        ECGenParameterSpec spec = new ECGenParameterSpec("secp256r1");
        kpg.initialize(spec);
        return kpg.generateKeyPair();
    }

    // ================================================================
    // Message Digests — Some Broken
    // ================================================================

    public static byte[] digestMD5(byte[] data) throws Exception {
        // MD5 — CLASSICALLY BROKEN
        MessageDigest md = MessageDigest.getInstance("MD5");
        return md.digest(data);
    }

    public static byte[] digestSHA1(byte[] data) throws Exception {
        // SHA-1 — CLASSICALLY BROKEN
        MessageDigest md = MessageDigest.getInstance("SHA-1");
        return md.digest(data);
    }

    public static byte[] digestSHA256(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        return md.digest(data);
    }

    // ================================================================
    // PBKDF2 — Key Derivation
    // ================================================================

    public static byte[] deriveKeyPBKDF2(char[] password, byte[] salt) throws Exception {
        // PBKDF2WithHmacSHA256 — KDF with SHA-256 HMAC
        SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        return skf.generateSecret(
            new javax.crypto.spec.PBEKeySpec(password, salt, 310000, 256)
        ).getEncoded();
    }
}
